Implemented segmented algorithms for partitioned vector

ajaivgeorge opened the pull request:
Tasks completed as part of GSoC project (See: #1338) -
  - Moved unit tests for segmented algorithms from component to
    parallel/segmented_algorithms
  - Implemented segmented for_each_n using existing segmented for_each
  - Implemented segmented unary transform
  - Implemented 2 versions of segmented binary transform with 4 and 5
    iterators as arguments
  - Implemented segmented reduce
  - Implemented segmented transform_inclusive/exclusive_scan by
    modifying existing inclusive/exculsive_scan to accept one more
    default parameter (unary operator) as input
  - modified dispatch.hpp in both algorithms/detail and
    segmented_algorithms/detail to add helper functions for output type
    inference and casting

All algorithms have associated passing tests in the unit/parallel/segmented_algorithms folder.
Inspect shows 0 errors
This branch has been updated with the almost latest hpx/master and only one test unrelated to my code fails (68 - tests.regressions.lcos_dir.wait_for_1751), which has been fixed in later commits.

hkaiser commented on hpx/include/parallel_for_each.hpp:
```diff
@@ -4,12 +4,11 @@
-#define HPX_PARALLEL_FOREACH_JUN_28_2014_0827AM
+#if !defined(HPX_PARALLEL_FOREACH_JUN_21_2017_0827AM)
+#define HPX_PARALLEL_FOREACH_JUN_21_2017_0827AM
 
```
Why this change?

mcopik replied:
I think it's simply a misunderstanding between me and the student. I told him to modify include guards but I meant only files he has created.

@ajaivgeorge could you revert those changes in files which have not been created by you?

hkaiser commented on hpx/include/parallel_reduce.hpp:
```diff
@@ -4,11 +4,11 @@
-#define HPX_PARALLEL_REDUCE_JUN_28_2014_0827AM
+#if !defined(HPX_PARALLEL_REDUCE_JUN_21_2017_0827AM)
+#define HPX_PARALLEL_REDUCE_JUN_21_2017_0827AM
 
```
Why this change?

hkaiser commented on hpx/parallel/algorithms/detail/dispatch.hpp:
```diff
@@ -18,7 +18,7 @@
 #include <hpx/parallel/util/detail/scoped_executor_parameters.hpp>
-
+#include <hpx/util/tuple.hpp>
 #include <string>
```
Please keep an empty line between the HPX headers and the std headers. This prevents clang-format from sorting those two groups of headers files into one block.

hkaiser commented on hpx/include/parallel_transform.hpp:
```diff
@@ -4,11 +4,10 @@
-#define HPX_PARALLEL_TRANSFORM_JUN_28_2014_0827AM
+#if !defined(HPX_PARALLEL_TRANSFORM_JUN_21_2017_0827AM)
+#define HPX_PARALLEL_TRANSFORM_JUN_21_2017_0827AM
 
```
There is no need to modify the include guards on every edit.

hkaiser commented on hpx/parallel/algorithms/for_each.hpp:
```diff
@@ -175,6 +192,37 @@ namespace hpx { namespace parallel { inline namespace v1
+            std::true_type, Proj && proj = Proj())
+        {
+            auto last = first;
+            detail::advance(last,std::size_t(count));
```
Please try to conform to our code formatting style. In particular we normally place a space after commas separating arguments.

hkaiser replied:
This comment applies to many spots in the code, btw. Clang-format will help with that as well (at least for the files you created).

hkaiser commented on hpx/parallel/algorithms/for_each.hpp:
```diff
@@ -463,7 +510,7 @@ namespace hpx { namespace parallel { inline namespace v1
     typename util::detail::algorithm_result<ExPolicy, InIter>::type
     for_each(ExPolicy && policy, InIter first, InIter last, F && f,
-        Proj && proj = Proj())
+        Proj && proj)
```
Why did you remove the default argument value?

ajaivgeorge replied:
for_each is forward declared above, so no default argument here.

hkaiser replied:
Ok, I missed that. Sorry.

hkaiser commented on hpx/parallel/algorithms/inclusive_scan.hpp:
```diff
@@ -184,11 +185,12 @@ namespace hpx { namespace parallel { inline namespace v1
+        inclusive_scan_(ExPolicy&& policy, InIter first, InIter last, OutIter
+            dest, T const& init, Op && op, std::true_type, Conv && conv =
+            Conv());
         /// \endcond
```
Please try not to separate argument type and argument name onto different lines.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/dispatch.hpp:
```diff
@@ -17,6 +17,8 @@
 #include <hpx/parallel/util/detail/algorithm_result.hpp>
 #include <hpx/parallel/util/detail/handle_remote_exceptions.hpp>
+#include <hpx/util/tuple.hpp>
+#include <exception>
```
This header is duplicated.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/scan.hpp:
```diff
@@ -43,15 +43,16 @@ namespace hpx { namespace parallel { inline namespace v1
+            Op && op, Conv && conv)
         {
-            T ret = *first;
+            T ret = conv(*first);
```
Please use `hpx::util::invoke` to call the predicates, etc.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/scan.hpp:
```diff
@@ -43,15 +43,16 @@ namespace hpx { namespace parallel { inline namespace v1
                 for(++first; first != last; ++first)
                 {
-                    ret = hpx::util::invoke(op, ret, *first);
+                    ret = hpx::util::invoke(op, ret, conv(*first));
```
Same here, please use `invoke`.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/scan.hpp:
```diff
@@ -95,9 +100,10 @@ namespace hpx { namespace parallel { inline namespace v1
                                 {
-                                    ret = hpx::util::invoke(op, ret, *curr);
+                                    ret = hpx::util::invoke(op, ret,
+                                        conv(*curr));
```
Same here, please use `invoke`.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/scan.hpp:
```diff
@@ -565,7 +573,7 @@ namespace hpx { namespace parallel { inline namespace v1
                     results.push_back(dispatch_async(traits_in::get_id(sit_in),
                         segmented_scan_T<T>(), policy,
-                        forced_seq(), beg, end, op));
+                        forced_seq(), beg, end, op, conv));
```
`conv` has to be perfectly forwarded (here and more places below).

hkaiser replied:
> `conv` has to be perfectly forwarded (here and more places below).

Sorry, I take that back ...

hkaiser commented on hpx/parallel/segmented_algorithms/reduce.hpp:
```diff
@@ -0,0 +1,252 @@
+#include <hpx/parallel/util/detail/algorithm_result.hpp>
+#include <hpx/parallel/util/detail/handle_remote_exceptions.hpp>
+
+#include <exception>
```
Please move this header into the std header block below (sorted).

hkaiser commented on hpx/parallel/segmented_algorithms/reduce.hpp:
```diff
@@ -0,0 +1,252 @@
+                            r.begin(), r.end(), init,
+                            [=](T const& val, shared_future<T>& curr)
+                            {
+                                return red_op(val, curr.get());
```
Shouldn't we use `hpx::util::invoke` here?

taeguk replied:
@hkaiser Is there a special reason to use ```hpx::util::invoke```?
When I implemented ```is_heap``` and ```is_heap_until```, I didn't use that.
Should I use it?
https://github.com/STEllAR-GROUP/hpx/blob/383eab90cad8a29dfd66f930d97bd50c77569689/hpx/parallel/algorithms/is_heap.hpp#L73

hkaiser replied:
I have already done this here: https://github.com/STEllAR-GROUP/hpx/blob/fixing_2699/hpx/parallel/algorithms/is_heap.hpp#L76.

The reason to use `invoke` that that it allows to transparently pass any callable to the algorithm, include member function pointers, in which case the current element is used as the `this` for the function invocation.

taeguk replied:
@hkaiser Awesome! I got it :)

taeguk replied:
@hkaiser Awesome! I got it :)
https://github.com/STEllAR-GROUP/hpx/blob/383eab90cad8a29dfd66f930d97bd50c77569689/hpx/parallel/algorithms/partition.hpp#L271

taeguk replied:
@hkaiser Awesome! I got it :)

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+            typename InIter2, typename OutIter, typename F,
+            typename Proj1, typename Proj2>
+        static typename util::detail::algorithm_result<ExPolicy,
+            hpx::util::tuple<InIter1,InIter2,OutIter>
```
Shouldn't this be a tagged_tuple instead?

ajaivgeorge replied:
@hkaiser @mcopik  At line 252 of segmented/transform, I convert the tuple to a tagged tuple? Is this enough? I based the implementation on the non segmented transform. So segmented_transform() returns a tuple and transform_() calls make_tagged_tuple on that and returns the result.

hkaiser commented on tests/unit/parallel/segmented_algorithms/partitioned_vector_for_each.cpp:
```diff
@@ -1,4 +1,4 @@
-//  Copyright (c) 2014-2015 Hartmut Kaiser
```
Why did you remove the copyright?

hkaiser commented on hpx/parallel/algorithms/exclusive_scan.hpp:
```diff
@@ -22,6 +22,7 @@
 #include <hpx/parallel/util/loop.hpp>
 #include <hpx/parallel/util/partitioner.hpp>
 #include <hpx/parallel/util/scan_partitioner.hpp>
+#include <hpx/parallel/util/projection_identity.hpp>
```
Please try to keep the headers sorted alphabetically (inside each of the blocks). Clang-format will do that for you.

hkaiser commented on hpx/parallel/algorithms/for_each.hpp:
```diff
@@ -278,15 +331,9 @@ namespace hpx { namespace parallel { inline namespace v1
-            std::forward<ExPolicy>(policy), is_seq(),
-            first, std::size_t(count), std::forward<F>(f),
-            std::forward<Proj>(proj));
+        return detail::for_each_n_(policy, first, count, f, is_segmented(),proj);
```
`policy`, `f`, and `proj` should be perfectly forwarded here.

hkaiser commented on hpx/parallel/algorithms/transform.hpp:
```diff
@@ -419,8 +447,8 @@ namespace hpx { namespace parallel { inline namespace v1
-                    auto f1 = transform_binary_iteration<ExPolicy, F, Proj1, Proj2>(
-                        std::forward<F>(f), std::forward<Proj1>(proj1),
+                    auto f1 = transform_binary_iteration<ExPolicy, F, Proj1,
+                        Proj2>(std::forward<F>(f), std::forward<Proj1>(proj1),
```
What has prompted you to reformat those lines?

hkaiser commented on hpx/parallel/algorithms/transform.hpp:
```diff
@@ -179,6 +180,37 @@ namespace hpx { namespace parallel { inline namespace v1
+            typedef std::integral_constant<bool,
+                    parallel::execution::is_sequenced_execution_policy<ExPolicy>
+                    ::value || !hpx::traits::is_forward_iterator<InIter>::value ||
+                   !hpx::traits::is_forward_iterator<OutIter>::value
```
The Boolean expression is very difficult to read thus way. Could you please try to keep every term separate, if possible, e.g.:

    parallel::execution::is_sequenced_execution_policy<ExPolicy>::value || 
    !hpx::traits::is_forward_iterator<InIter>::value || 
    !hpx::traits::is_forward_iterator<OutIter>::value

mcopik commented on tests/unit/parallel/segmented_algorithms/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+set(partitioned_vector_transform_reduce_PARAMETERS
+    LOCALITIES 2
+    THREADS_PER_LOCALITY 4)
+set(partitioned_vector_reduce_FLAGS DEPENDENCIES partitioned_vector_component)
```
Is there a specific reason why `reduce` test is not executed on multiple localities?

ajaivgeorge replied:
This has been fixed.

hkaiser commented on hpx/parallel/algorithms/transform.hpp:
```diff
@@ -541,11 +613,15 @@ namespace hpx { namespace parallel { inline namespace v1
         traits::is_projected<Proj1, InIter1>::value &&
-        traits::is_projected<Proj2, InIter2>::value &&
+        traits::is_projected<Proj2, InIter2>::value)
+#if defined(HPX_MSVC) && HPX_MSVC <= 1800     // MSVC12 can't pattern match this
```
Where does this come from? The comment disagrees with the condition...

ajaivgeorge replied:
I have mistakenly put the MSVC check instead of the NVCC check here.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/dispatch.hpp:
```diff
@@ -145,6 +178,49 @@ namespace hpx { namespace parallel { inline namespace v1 { namespace detail
+                        typename traits3::local_iterator
+                    >
+                {
+                    auto p = f.get();
```
This could be 

    auto && p = f.get();

which would avoid a copy for non-move-constructable types.

ajaivgeorge replied:
should I update this for the other algorithm result helpers?

hkaiser replied:
Please go ahead with this if you find spots where this applies.

ajaivgeorge replied:
I reverted this change while trying to figure out the errors with transform. I will make this change (auto p to auto && p) on this and other helpers again.

hkaiser commented on hpx/parallel/segmented_algorithms/exclusive_scan.hpp:
```diff
@@ -117,25 +117,27 @@ namespace hpx { namespace parallel { inline namespace v1
         segmented_exclusive_scan_seq(ExPolicy && policy, SegIter first,
-            SegIter last, OutIter dest, T const& init, Op && op, std::false_type)
+            SegIter last, OutIter dest, T const& init, Op && op,
+            std::false_type, Conv && conv)
```
`conv` is not used in this function, is this intended?

ajaivgeorge replied:
This and the below places where Conv is not used are for the non segmented OutIter implementation of segmented InIter inclusive/exclusive scan. @mcopik told me to ignore that for now, but i kept the function signature the same as the other std::true_type template of these functions, which are for the segmented OutIter version.

hkaiser replied:
Ok. If the parameter is not needed we might want to remove it altogether to avoid confusing people.

ajaivgeorge replied:
Please look at line 215, I think conv will have to be passed to both templated versions, or we will need to separate them out.

hkaiser commented on hpx/parallel/segmented_algorithms/exclusive_scan.hpp:
```diff
@@ -157,26 +159,28 @@ namespace hpx { namespace parallel { inline namespace v1
         segmented_exclusive_scan_par(ExPolicy && policy, SegIter first,
-            SegIter last, OutIter dest, T const& init, Op && op, std::false_type)
+            SegIter last, OutIter dest, T const& init, Op && op,
+            std::false_type, Conv && conv)
```
Same here, `conv` is not used.

hkaiser commented on hpx/parallel/segmented_algorithms/inclusive_scan.hpp:
```diff
@@ -113,26 +113,28 @@ namespace hpx { namespace parallel { inline namespace v1
         segmented_inclusive_scan_seq(ExPolicy && policy, SegIter first,
-            SegIter last, OutIter dest, T const& init, Op && op, std::false_type)
+            SegIter last, OutIter dest, T const& init, Op && op,
+            std::false_type, Conv && conv)
```
Same here, `conv` is unused.

hkaiser commented on hpx/parallel/segmented_algorithms/inclusive_scan.hpp:
```diff
@@ -154,27 +156,29 @@ namespace hpx { namespace parallel { inline namespace v1
         segmented_inclusive_scan_par(ExPolicy && policy, SegIter first,
-            SegIter last, OutIter dest, T const& init, Op && op, std::false_type)
+            SegIter last, OutIter dest, T const& init, Op && op,
+            std::false_type, Conv && conv)
```
unused `conv`, as above - is that intended?

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+                typename iterator_traits1::local_iterator,
+                typename iterator_traits2::local_iterator> >(),
+                std::forward<ExPolicy>(policy), first, last,
+                dest, std::forward<F>(f), proj, is_seq()));
```
Please add proper indentation to make this snippet more readable. Also, `proj` should be perfectly forwarded:

            return hpx::util::make_tagged_pair<tag::in, tag::out>( 
                 segmented_transform(
                     transform<std::pair<
                         typename iterator_traits1::local_iterator, 
                         typename iterator_traits2::local_iterator
                     > >(), 
                     std::forward<ExPolicy>(policy), first, last, 
                     dest, std::forward<F>(f), std::forward<Proj>(proj), 
                     is_seq()));

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+                hpx::util::tuple<InIter1,InIter2,OutIter>
+            > result;
+            auto last2 = first2;
+            detail::advance(last2,std::size_t(last1 - first1));
```
Are `last1` and `first1` guaranteed to be random access iterator?

ajaivgeorge replied:
Should I restrict the iterators to be random iterators using a check or is there a better more generic way to find the last (end) iterator given the first (begin) iterator?

mcopik replied:
@ajaivgeorge No, we should accept all types of iterators. `std::distance` gives you a distance between two iterators.

As you can see [here](http://en.cppreference.com/w/cpp/iterator), only random access iterators have a subtraction operator. If you want to learn more, Hartmut gave last year [a series of excellent lectures](https://www.youtube.com/playlist?list=PL7vEgTL3FalY2eBxud1wsfz8OKvE9sd_z) on STL and template metaprogramming. The first one gives an introduction to different types of iterators. In the second lecture, the `advance` function is analyzed.

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+            segment_iterator3 sdest =  traits3::segment(dest);
+
+            std::vector<future<hpx::util::tuple<local_iterator_type1,
+             local_iterator_type2, local_iterator_type3> > > segments;
```
Indentation gone wrong?

mcopik replied:
@ajaivgeorge Can you improve the indentation here? A typedef for tuple may help here, you could reuse it at line 479.

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+                }
+            }
+
+            return result::get((
```
Why '(( ... '))'?

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+            typedef util::detail::algorithm_result<ExPolicy,
+            hpx::util::tagged_tuple<
+                tag::in1(InIter1), tag::in2(InIter2), tag::out(OutIter)>
+            > result;
```
Is `result` unused?

ajaivgeorge replied:
Used in 517

hkaiser replied:
Ahh, the indentation is confusing, I see this now.

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+                tag::in1(InIter1), tag::in2(InIter2), tag::out(OutIter)>
+            > result;
+            auto last2 = first2;
+            detail::advance(last2,std::size_t(last1 - first1));
```
Are `last1` and `first1` guaranteed to be random access iterators?

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+                typename iterator2_traits::local_iterator,
+                typename iterator3_traits::local_iterator> >(),
+                std::forward<ExPolicy>(policy), first1, last1, first2,
+                dest, std::forward<F>(f), proj1, proj2, is_seq()));
```
`proj1` and `proj2` should be perfectly forwarded. Please fix formatting (see above).

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+                }
+
+                // handle all of the full partitions
+                if(last1-first1 <= last2-first2)
```
Random access iterators?

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+            segment_iterator3 sdest =  traits3::segment(dest);
+
+            std::vector<future<hpx::util::tuple<local_iterator_type1,
+             local_iterator_type2, local_iterator_type3> > > segments;
```
Indentation gone wrong?

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+                }
+
+                // handle all of the full partitions
+                if(last1-first1 <= last2-first2)
```
Random access iterators?

hkaiser commented on hpx/parallel/segmented_algorithms/transform.hpp:
```diff
@@ -0,0 +1,876 @@
+            typedef util::detail::algorithm_result<ExPolicy,
+            hpx::util::tagged_tuple<
+                tag::in1(InIter1), tag::in2(InIter2), tag::out(OutIter)>
+            > result;
```
`result` is unused in this context.

ajaivgeorge replied:
It is used in line 841.

hkaiser replied:
Same here, strange indentation...

mcopik replied:
The indentation is not consistent here, e.g. why the indentation style for `tagged_tuple` is different in lines 820 and 833?

ajaivgeorge commented:
@hkaiser , @mcopik please have a look and see if this PR is ready for merging.

mcopik commented:
@ajaivgeorge Have you fixed the issue with ctest and verified that all tests are passing on multiple localities?

ajaivgeorge commented:
@mcopik, yes. I have modified the CMakeLists file in the segmented_algorithms test folder.

ajaivgeorge commented:
@mcopik, Only, transform fails on rostam. I forgot that CI wasn't running the tests. I will fix this first and then get back to find_end/first_of. Are you sure find test fails? I ran it on rostam and there was no SegFault. Find end failed as expected.

ajaivgeorge commented:
@mcopik Errors have been fixed. Please review.

hkaiser commented:
Is this ready to be reviewed now?

ajaivgeorge commented:
@hkaiser Yes it is ready to be reviewed.

hkaiser commented on hpx/parallel/algorithms/exclusive_scan.hpp:
```diff
@@ -6,8 +6,8 @@
-#if !defined(HPX_PARALLEL_ALGORITHM_EXCLUSIVE_SCAN_DEC_30_2014_1236PM)
-#define HPX_PARALLEL_ALGORITHM_EXCLUSIVE_SCAN_DEC_30_2014_1236PM
+#if !defined(HPX_PARALLEL_ALGORITHM_EXCLUSIVE_SCAN_JUN_21_2017_1236PM)
+#define HPX_PARALLEL_ALGORITHM_EXCLUSIVE_SCAN_JUN_21_2017_1236PM
```
There is no need to change the include guards.

hkaiser commented on hpx/parallel/algorithms/exclusive_scan.hpp:
```diff
@@ -146,16 +150,16 @@ namespace hpx { namespace parallel { inline namespace v1
+                    [op, conv](zip_iterator part_begin, std::size_t part_size) -> T
                     {
-                        T part_init = get<0>(*part_begin++);
+                        T part_init = hpx::util::invoke(conv, get<0>(*part_begin++));
```
This additionally requires a check whether `part_begin` has not reached the end after the increment.

hkaiser replied:
I think it should return `part_init`.

hkaiser replied:
> I looked at an example from algorithms/reduce.hpp L72, and even here there is no such check.

There it is not a problem as `part_size` would drop to zero, thus avoiding to dereference the iterator produced by `++part_begin` inside `accumulate_n`.

hkaiser commented on hpx/parallel/algorithms/inclusive_scan.hpp:
```diff
@@ -6,8 +6,8 @@
-#if !defined(HPX_PARALLEL_ALGORITHM_INCLUSIVE_SCAN_JAN_03_2015_0136PM)
-#define HPX_PARALLEL_ALGORITHM_INCLUSIVE_SCAN_JAN_03_2015_0136PM
+#if !defined(HPX_PARALLEL_ALGORITHM_INCLUSIVE_SCAN_JUN_21_2017_0136PM)
+#define HPX_PARALLEL_ALGORITHM_INCLUSIVE_SCAN_JUN_21_2017_0136PM
```
There is no need to change the include guards.

hkaiser commented on hpx/parallel/algorithms/inclusive_scan.hpp:
```diff
@@ -140,16 +144,16 @@ namespace hpx { namespace parallel { inline namespace v1
                     {
-                        T part_init = get<0>(*part_begin);
+                        T part_init = hpx::util::invoke(conv, get<0>(*part_begin));
                         get<1>(*part_begin++) = part_init;
```
This requires an additional check that `part_begin` has not reached the end after the  increment.

hkaiser commented on hpx/parallel/algorithms/transform.hpp:
```diff
@@ -180,6 +181,55 @@ namespace hpx { namespace parallel { inline namespace v1
+                (hpx::traits::is_forward_iterator<FwdIter1>::value),
+                "Requires at least forward iterator.");
+            static_assert(
+                (hpx::traits::is_input_iterator<FwdIter2>::value),
```
Shouldn't this check for `forward_iterator_tag` instead?

hkaiser commented on hpx/parallel/segmented_algorithms/detail/dispatch.hpp:
```diff
@@ -3,15 +3,16 @@
-#if !defined(HPX_PARALLEL_ALGORITHM_REMOTE_DISPATCH_OCT_15_2014_0938PM)
-#define HPX_PARALLEL_ALGORITHM_REMOTE_DISPATCH_OCT_15_2014_0938PM
+#if !defined(HPX_PARALLEL_ALGORITHM_REMOTE_DISPATCH_JUN_21_2017_0938PM)
+#define HPX_PARALLEL_ALGORITHM_REMOTE_DISPATCH_JUN_21_2017_0938PM
```
There is no need to change the include guards.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/reduce.hpp:
```diff
@@ -0,0 +1,125 @@
+                    return util::accumulate_n(++part_begin, --part_size,
+                        std::move(val), r);
+                },
+                hpx::util::unwrapped([r](std::vector<T> && results)
```
Please add explicit return types to all lambdas.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/reduce.hpp:
```diff
@@ -0,0 +1,125 @@
+            typedef typename std::iterator_traits<InIter>::reference
+                reference;
+            return util::accumulate<T>(first, last,
+                [=](T const& res, reference next)
```
Same here, please add an explicit return type to the lambda.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/reduce.hpp:
```diff
@@ -0,0 +1,125 @@
+                        std::move(val),
+                        // MSVC14 bails out if r and conv are captured by
+                        // reference
+                        [=](T const& res, reference next)
```
Explicit return type missing... Same applies to code below.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/reduce.hpp:
```diff
@@ -0,0 +1,125 @@
+                },
+                hpx::util::unwrapped([r](std::vector<T> && results)
+                {
+                    auto rfirst = std::begin(results);
```
Please use `hpx::util::begin`/`hpx::util::end` instead of the standard versions.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/scan.hpp:
```diff
@@ -637,14 +651,14 @@ namespace hpx { namespace parallel { inline namespace v1
                         policy.executor(),
                         hpx::util::unwrapped(
-                            [=, &op](T last_value, T)
+                            [=, &op, &conv](T last_value, T)
```
Please add an explicit return type to this lambda.

ajaivgeorge replied:
The return of segmented_scan_void function object is hpx::util::unused. SHould this be the return of he lambda or should it be blank for void?

hkaiser replied:
The lambda returns `void`.

hkaiser commented on hpx/parallel/util/loop.hpp:
```diff
@@ -539,6 +540,22 @@ namespace hpx { namespace parallel { namespace util
+        {
+            val = hpx::util::invoke(r, val, *iter);
+            iter++;
+        };
```
Stray semicolon, also - why don't you use `first` for the iteration?

hkaiser commented:
@ajaivgeorge Overall, very nice job - thanks for your effort!

mcopik commented:
@ajaivgeorge I don't see any other problems besides those mentioned by Hartmut already. For the scan, it makes sense to verify if `part_begin` is not equal to end after an incrementation. Indentation in typedefs should be consistent and it's not, e.g. compare lines 818, 829 and 832 in transform.hpp; sometimes you use indentation for template parameters and sometimes not, sometimes the indentation is used in the last line and sometimes not.

@hkaiser can we enforce a CI build on AppVeyor? It seems it failed due to a network problem.

hkaiser commented:
> can we enforce a CI build on AppVeyor? It seems it failed due to a network problem.

AppVeyor seems to have problems to pull from repository forks... I can try rebuilding, though.

ajaivgeorge commented:
@hkaiser, @mcopik. I have improved the indentation in segmented/transform.hpp and have added the required checks to the part_begin in the scans. Do let me know if any more changes are required.

hkaiser commented:
@mcopik: are you fine with going ahead and merging this?

mcopik commented:
@hkaiser Yes, let's do it :)

hkaiser commented:
@ajaivgeorge, @mcopik: two of the new tests are failing in distributed runs, see here for an example: http://rostam.cct.lsu.edu/builders/hpx_gcc_4_9_4_boost_1_56_centos_x86_64_release/builds/126/steps/run_unit_tests/logs/stdio

Could you fix those asap, please?

aserio commented:
@ajaivgeorge, congratulations on your commit to HPX. Once you get the Buildbot issues sorted out, contact me so that we can send you a T-Shirt

ajaivgeorge commented:
@hkaiser @mcopik, The error has been fixed in my master branch. The issue was with the util/loop accumulate function that I added. During the PR review when i changed from using a temp iterator to using first iterator directly i forgot to add a ++first. Hadn't run the test after this change.

I have tested the algorithm upto localities=5. The other algorithms like transform_scan test and for_each and transform also work for localities > 2. But the following tests fail for localities > 2
1) handle_values
2) iter
3) fill
4) inclusive/exclusive_scan - I though that this might be because of the changes I made but the transform_scan test works so I am not sure.

These might be failing because the tests are designed for localities=2 (especially the extensive tests for the scans) or because of something fundamental with the algorithm. I will look into this once I am done with the find algorithms.

hkaiser commented:
@ajaivgeorge: please create a new PR for the algorithm fixes.
